### ISSUE TITLE
Compile cscs to exe?

### ISSUE THREAD

**Issue by djlw78** [vassilych/cscs#19]
How would compiling to exe look/work?

**Comment by vassilych** [comment]
You mean on Windows? Yes, it's compiled under bin/Debug for instance and
called cscs.exe
You can submit any CSCS script as a command-line argument to that exe, e.g.
cscs.exe myscript.cscs
Actually almost same thing happens on a Mac.

**Comment by djlw78** [comment]
Oh no I mean MyScript.exe 
Compiling scripts or script clusters to an portable executable file.

**Comment by djlw78** [comment]
https://www.codemag.com/Article/2001071 but there is no perfect example on precompiler.

**Comment by vassilych** [comment]
Ok, I see. Then you'd need to do a small change in Main.cs. Instead of
reading the scripting file from the command-line, you "hardcode" the
filename in the code like this (ca. line 40):

            string scriptFilename = "somepath/myscript.cscs";
            string script = Utils.GetFileContents(scriptFilename);

But the script must under somepath at runtime (could be in the same
directory as the binary itself).
You might have a question if you can encode the contents of the script so
that if you distribute the executable together with the script the clients
won't be able to see the scripting code.
If this is the question, I am currently working on it.. Let me know...

**Comment by djlw78** [comment]
Yeah, that's almost useless I ripped this code out "DebuggerServer.StartServer(13337, true);" by commenting itself and the if statement around it because it was generating a very disliked outcome for me, after all of that was removed it ran fine. (I don't wish to connect to the remote debugger every run), still unsure about how to compile the cscs script to and exe file with includes, because there is no documentation on precompiler because the arguments for it are "string functionName, string[] args, Dictionary<string, Variable> argsMap, string cscsCode, ParsingScript parentScript" and I am not even sure if the precompiler is the thing I am looking towards to compile.

**Comment by vassilych** [comment]
What exactly is almost useless?
First step to include a script into the binary is to add those two lines
(or edit them) that I mentioned in my previous e-mail:

string scriptFilename = "somepath/myscript.cscs";
string script = Utils.GetFileContents(scriptFilename);

After this work next step is the precompiled stuff. I didn't get if this
step worked for you or not.
Obviously you should remove all the stuff you don't need, like the Debugger
(it is used to connect from VS Code actually).

**Comment by djlw78** [comment]
Yes that part works fine,  not sure what to fill in for precompiler(args...) though.

**Comment by vassilych** [comment]
Not sure I got your question - args where, in what call?

**Comment by djlw78** [comment]
Its just not documented how to tell the cscs compile to convert cscs files into exe files, or dlls for some matter (dll is less important) I think the pre-compiler is in charge of that but I am not sure how to use it.

**Comment by vassilych** [comment]
No, you don't compile it directly into an .exe file. When Pre-compiler
compiles a function, it puts it into a dll. But this dll is created at
runtime and added to the executing binary either at start-up, when a
start-up script contains a pre-compiled function (the keyword is
"cfunction" for such function) or any time after the start-up...

**Comment by djlw78** [comment]
Okay I am getting no where will this i'll just find a way to compile it exe myself.

**Comment by vassilych** [comment]
Can you let me know if you have a particular question or if there are
questions to anything I wrote before?
